modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/ForEachProcessor.java

```diff
@@ -79,41 +76,35 @@ public void execute(IngestDocument ingestDocument, BiConsumer<IngestDocument, Ex
                 handler.accept(null, new IllegalArgumentException("field [" + field + "] is null, cannot loop over its elements."));
             }
         } else {
-            List<Object> newValues = new CopyOnWriteArrayList<>();
-            innerExecute(0, values, newValues, ingestDocument, handler);
+            innerExecute(0, values, Collections.synchronizedList(new ArrayList<>(values.size())), ingestDocument, handler);
         }
     }
 
     void innerExecute(int index, List<?> values, List<Object> newValues, IngestDocument document,
                       BiConsumer<IngestDocument, Exception> handler) {
+        for (; index < values.size(); index++) {
+            AtomicBoolean shouldContinueHere = new AtomicBoolean();
```

Comment:
I like this approach. In the case of inner processor going async; whichever thread sets this atomic boolean last should continue with the next value.

And in the case inner processor doesn't go async then on the second `getAndSet(...)` the current thread continues with the next value.

Do wonder if this variable can be renamed to describe this beter.

Reply:
I wonder about this scenario:
1) The current thread sets `shouldContinueHere` first.
2) Next value is handled  and also manages to set `shouldContinueHere` first.
3) Both async call manage to return and update `newValues` simultaneously.

Is this scenario possible? And if so should have protection against this?

Reply:
If current thread hits `shouldContinueHere` first, then it sees `false` and ends the loop (returns). 
If it hits it second that means it's after async call already finished and other thread updated `newValues` already.
As you mentioned in first comment only thread that checks `shouldContinueHere` last actually proceeds
Does it answer your question or did I misunderstand the scenario?

Reply:
Thanks for clarifying. I misinterpreted the if statement on line 98 when looking at it this morning.
It is not possible that multiple values from the list handled concurrently.